components/gitpod-cli/cmd/validate.go

```diff
@@ -487,11 +503,102 @@ func openWindow(ctx context.Context, workspaceUrl string) error {
 	return gpCmd.Run()
 }
 
+func waitForAllTasksToOpen(ctx context.Context, supervisor *supervisor.SupervisorClient, runLog *logrus.Entry) (tasks []*api.TaskStatus, allTasksOpened bool) {
+	for !allTasksOpened {
+		time.Sleep(1 * time.Second)
+		if ctx.Err() != nil {
+			return
+		}
+		listener, err := supervisor.Status.TasksStatus(ctx, &api.TasksStatusRequest{
+			Observe: true,
+		})
+		if err != nil {
+			continue
+		}
+		tasks, allTasksOpened = checkAllTasksOpened(ctx, listener, runLog)
+	}
+	return
+}
+
+func checkAllTasksOpened(ctx context.Context, listener api.StatusService_TasksStatusClient, runLog *logrus.Entry) (tasks []*api.TaskStatus, allTasksOpened bool) {
+	for !allTasksOpened {
+		resp, err := listener.Recv()
+		if err != nil {
+			return
+		}
+		tasks = resp.GetTasks()
+		allTasksOpened = areTasksOpened(tasks)
+	}
+	return
+}
+
+func areTasksOpened(tasks []*api.TaskStatus) bool {
+	for _, task := range tasks {
+		if task.State == api.TaskState_opening {
+			return false
+		}
+	}
+	return true
+}
+
+func pipeTask(ctx context.Context, task *api.TaskStatus, supervisor *supervisor.SupervisorClient, runLog *logrus.Entry) {
+	taskTitle := task.Presentation.Name
+
+	listen, err := supervisor.Terminal.Listen(ctx, &api.ListenTerminalRequest{
+		Alias: task.Terminal,
+	})
+	if ctx.Err() != nil {
+		return
+	}
+	if err != nil {
+		runLog.WithError(err).Errorf("%s: failed to attach to terminal", taskTitle)
+		return
+	}
+
+	var buffer bytes.Buffer
+	defer func() {
+		if buffer.Len() > 0 {
+			runLog.Infof("%s: %s", taskTitle, buffer.String())
+		}
+	}()
+	for {
+		resp, err := listen.Recv()
+		if err == io.EOF || ctx.Err() != nil {
+			return
+		}
+		if err != nil {
+			runLog.WithError(err).Errorf("%s: failed to receive data", taskTitle)
+			return
+		}
+
+		title := resp.GetTitle()
+		if title != "" {
+			taskTitle = title
+		}
+
+		exitCode := resp.GetExitCode()
+		if exitCode != 0 {
+			runLog.Infof("%s: exited with code %d", taskTitle, exitCode)
+		}
+
+		data := resp.GetData()
+		if len(data) > 0 {
+			buffer.Write(data)
+			for bytes.Contains(buffer.Bytes(), []byte("\n")) {
```

Comment:
@iQQBot it works, but i wonder is there nicer way, I tried with Scanner to scan lines, but it just hangs

Reply:
let me try, if not could be done in another PR?

Reply:
thx @iQQBot io.Pipe did a trick, I also added retry loop for reliability

could you have a look once again please?